File: python/Q235_lowest_common_ancestor_of_a_binary_search_tree.py
#
# @lc app=leetcode id=235 lang=python3
#
# [235] Lowest Common Ancestor of a Binary Search Tree
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x: int) -> None:
        self.val: int = x
        self.left: TreeNode | None = None
        self.right: TreeNode | None = None


# @lc code=start
class Solution:
    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        if root.val > p.val and root.val > q.val:
            assert root.left is not None
            return self.lowestCommonAncestor(root.left, p, q)

        if root.val < p.val and root.val < q.val:
            assert root.right is not None
            return self.lowestCommonAncestor(root.right, p, q)

        return root


# @lc code=end

File: rust/src/leetcode/Q235_lowest_common_ancestor_of_a_binary_search_tree.rs
/*
 * @lc app=leetcode id=235 lang=rust
 *
 * [235] Lowest Common Ancestor of a Binary Search Tree
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    fn lowest_common_ancestor_in_range(
        root: Option<Rc<RefCell<TreeNode>>>,
        left: i32,
        right: i32,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if let Some(node) = root {
            if node.borrow().val < left {
                Self::lowest_common_ancestor_in_range(node.borrow().right.clone(), left, right)
            } else if node.borrow().val > right {
                Self::lowest_common_ancestor_in_range(node.borrow().left.clone(), left, right)
            } else {
                Some(node)
            }
        } else {
            None
        }
    }

    pub fn lowest_common_ancestor(
        root: Option<Rc<RefCell<TreeNode>>>,
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        let p_value: i32 = p.unwrap().borrow().val;
        let q_value: i32 = q.unwrap().borrow().val;

        Self::lowest_common_ancestor_in_range(
            root.clone(),
            p_value.min(q_value),
            p_value.max(q_value),
        )
    }
}
// @lc code=end
